File: presentations/14-repository-and-builder-pattern-demos/csharp-smart-home-scenes/Domain/DeviceScene.cs
namespace SmartHomeScenesDemo.Domain;

public sealed class DeviceScene
{
    public DeviceScene(Guid id, string name, IReadOnlyList<SceneAction> actions)
    {
        if (id == Guid.Empty)
        {
            throw new ArgumentException("Scene id must not be empty.", nameof(id));
        }

        if (string.IsNullOrWhiteSpace(name))
        {
            throw new ArgumentException("Scene name is required.", nameof(name));
        }

        if (actions.Count == 0)
        {
            throw new ArgumentException("A scene must contain at least one action.", nameof(actions));
        }

        Id = id;
        Name = name;
        Actions = actions
            .OrderBy(action => action.OrderIndex)
            .ToArray();
    }

    public Guid Id { get; }

    public string Name { get; }

    public IReadOnlyList<SceneAction> Actions { get; }
}

File: presentations/14-repository-and-builder-pattern-demos/java-smart-home-scenes/src/main/java/edu/ksu/swe4743/smarthomescenes/domain/DeviceScene.java
package edu.ksu.swe4743.smarthomescenes.domain;

import java.util.Comparator;
import java.util.List;
import java.util.Objects;
import java.util.UUID;

public final class DeviceScene {
    private final UUID id;
    private final String name;
    private final List<SceneAction> actions;

    public DeviceScene(UUID id, String name, List<SceneAction> actions) {
        this.id = Objects.requireNonNull(id, "id");
        this.name = Objects.requireNonNull(name, "name").trim();

        if (this.name.isBlank()) {
            throw new IllegalArgumentException("Scene name is required.");
        }

        if (actions.isEmpty()) {
            throw new IllegalArgumentException("A scene must contain at least one action.");
        }

        this.actions = actions.stream()
            .sorted(Comparator.comparingInt(SceneAction::getOrderIndex))
            .toList();
    }

    public UUID getId() {
        return id;
    }

    public String getName() {
        return name;
    }

    public List<SceneAction> getActions() {
        return actions;
    }
}
